File: androidApp/src/main/kotlin/app/cycluna/android/features/journal/MoodCharts.kt
package app.cycluna.android.features.journal

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.cycluna.android.LocalCycleStore
import app.cycluna.android.core.CycleStore
import app.cycluna.android.core.MoodScale
import app.cycluna.android.designsystem.Theme
import app.cycluna.android.features.home.MoonDisc
import app.cycluna.android.features.phases.PhaseContent
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale
import kotlin.math.cos

/**
 * The three chart bodies behind "Mood patterns". Each takes one prepared page and draws it —
 * no aggregation here, that all lives in the shared core.
 *
 * Shared conventions: y is mood 1..5 (bottom to top), a dot per real log, and a connecting
 * line only once there are enough points for the shape to mean something.
 */
private object Plot {
    /**
     * Wide enough that a dot on the first or last position isn't clipped by the Canvas edge.
     * At 6dp a day-1 log was drawn half outside the chart and read as missing.
     */
    val PAD_X = 12.dp
    val PAD_TOP = 12.dp
    val PAD_BOTTOM = 10.dp

    /** Below this, joining dots implies a trend the data doesn't show. */
    const val MIN_POINTS_FOR_LINE = 4

    fun DrawScope.yOf(mood: Double): Float {
        val h = size.height - PAD_TOP.toPx() - PAD_BOTTOM.toPx()
        return PAD_TOP.toPx() + ((5 - mood) / 4).toFloat() * h
    }

    /**
     * One mood dot, ringed in the card colour so overlapping dots stay countable —
     * consecutive days sit only ~11dp apart on the moon lens, closer than a dot is wide.
     */
    fun DrawScope.moodDot(at: Offset, color: Color) {
        drawCircle(Theme.surface, 6.5.dp.toPx(), at)
        drawCircle(color, 5.dp.toPx(), at)
    }

    fun DrawScope.connect(points: List<Offset>, color: Color) {
        if (points.size < MIN_POINTS_FOR_LINE) return
        val path = Path().apply {
            points.forEachIndexed { i, p -> if (i == 0) moveTo(p.x, p.y) else lineTo(p.x, p.y) }
        }
        drawPath(
            path,
            color,
            style = Stroke(2.5.dp.toPx(), cap = StrokeCap.Round, join = StrokeJoin.Round),
        )
    }
}

private val DAY_OF_MONTH = DateTimeFormatter.ofPattern("d", Locale.ENGLISH)
private val DAY_MONTH = DateTimeFormatter.ofPattern("d MMM", Locale.ENGLISH)

private fun parse(iso: String): LocalDate? = runCatching { LocalDate.parse(iso) }.getOrNull()

// ------------------------------------------------------------------------------------
// Daily
// ------------------------------------------------------------------------------------

@Composable
fun DailyMoodChart(page: CycleStore.MoodPage, modifier: Modifier = Modifier) {
    val start = parse(page.startIso)
    Canvas(modifier.clearAndSetSemantics {}) {
        val w = size.width - Plot.PAD_X.toPx() * 2
        val span = (page.spanDays - 1).coerceAtLeast(1)
        fun xOf(dayOffset: Int) = Plot.PAD_X.toPx() + (dayOffset.toFloat() / span) * w

        // Faint gridlines at each mood level, so height is readable without an axis.
        with(Plot) {
            for (level in 1..5) {
                val ly = yOf(level.toDouble())
                drawLine(
                    Theme.inkSoft.copy(alpha = if (level == 3) 0.14f else 0.07f),
                    Offset(Plot.PAD_X.toPx(), ly),
                    Offset(size.width - Plot.PAD_X.toPx(), ly),
                    strokeWidth = 1.dp.toPx(),
                )
            }

            val points = page.daily.mapNotNull { d ->
                val date = parse(d.dateIso) ?: return@mapNotNull null
                val offset = start?.let { ChronoUnit.DAYS.between(it, date).toInt() } ?: return@mapNotNull null
                Offset(xOf(offset), yOf(d.mood.toDouble())) to d.mood
            }
            connect(points.map { it.first }, Theme.primary.copy(alpha = 0.55f))
            points.forEach { (p, mood) -> moodDot(p, MoodScale.color(mood)) }
        }
    }
}

@Composable
fun DailyMoodAxis(page: CycleStore.MoodPage, modifier: Modifier = Modifier) {
    val start = parse(page.startIso)
    Row(
        modifier
            .fillMaxWidth()
            .padding(horizontal = Plot.PAD_X)
            .semantics { contentDescription = "${page.title}, ${page.summary.count} logs" },
    ) {
        (0 until page.spanDays).forEach { offset ->
            Text(
                // Only every other day, or fourteen labels collide.
                if (offset % 2 == 0) start?.plusDays(offset.toLong())?.format(DAY_OF_MONTH) ?: "" else " ",
                Modifier.weight(1f),
                fontSize = 8.sp,
                color = Theme.inkSoft,
                textAlign = TextAlign.Center,
            )
        }
    }
}

// ------------------------------------------------------------------------------------
// Phase
// ------------------------------------------------------------------------------------

@Composable
fun PhaseMoodChart(page: CycleStore.MoodPage, periodLength: Int, modifier: Modifier = Modifier) {
    Canvas(modifier.clearAndSetSemantics {}) {
        val w = size.width - Plot.PAD_X.toPx() * 2
        val h = size.height - Plot.PAD_TOP.toPx() - Plot.PAD_BOTTOM.toPx()
        val length = page.spanDays.coerceAtLeast(1)
        fun xOf(day: Int) =
            Plot.PAD_X.toPx() + ((day - 1).toFloat() / (length - 1).coerceAtLeast(1)) * w

        // Phase bands sized to THIS cycle's real length, not the predicted one.
        PhaseContent.entries.forEach { phase ->
            val r = phase.dayRange(length, periodLength)
            val x0 = xOf(r.first)
            val x1 = xOf(minOf(r.last + 1, length))
            if (x1 > x0) {
                drawRect(
                    phase.color.copy(alpha = 0.12f),
                    Offset(x0, Plot.PAD_TOP.toPx()),
                    Size(x1 - x0, h),
                )
            }
        }

        with(Plot) {
            val points = page.cycle.map { Offset(xOf(it.cycleDay), yOf(it.mood.toDouble())) to it.mood }
            connect(points.map { it.first }, Theme.primary.copy(alpha = 0.6f))
            points.forEach { (p, mood) -> moodDot(p, MoodScale.color(mood)) }
        }
    }
}

/** Roughly this many date ticks across the cycle — enough to locate yourself, few enough
 *  that they don't collide on a narrow screen. */
private const val DATE_TICKS = 5

@Composable
fun PhaseMoodAxis(page: CycleStore.MoodPage, periodLength: Int, modifier: Modifier = Modifier) {
    val span = page.spanDays.coerceAtLeast(1)
    val start = parse(page.startIso)
    val step = (span / (DATE_TICKS - 1)).coerceAtLeast(1)
    val tickDays = (1..span step step).toList()

    Column(
        modifier
            .fillMaxWidth()
            .semantics {
                contentDescription = "${page.title}, ${page.spanDays} days, ${page.summary.count} logs"
            },
        verticalArrangement = Arrangement.spacedBy(2.dp),
    ) {
        // Both rows size to their own text. Pinning them to 13dp and 12dp clipped the
        // descenders off "Menstrual" and "28 Jun" — and would have sliced the labels in half
        // for anyone running a larger font scale, where the text grows but a fixed dp box
        // does not. `lineHeight` keeps the two rows tight without cropping them.
        Row(Modifier.fillMaxWidth()) {
            PhaseContent.entries.forEach { phase ->
                val r = phase.dayRange(span, periodLength)
                val days = (r.last - r.first + 1).coerceAtLeast(0)
                if (days > 0) {
                    Text(
                        phase.key,
                        Modifier.weight(days.toFloat()),
                        fontSize = 9.sp,
                        lineHeight = 11.sp,
                        color = Theme.inkSoft,
                        maxLines = 1,
                        textAlign = TextAlign.Center,
                    )
                }
            }
        }

        // Real calendar dates as well as cycle phases — "day 14" is hard to place in a month
        // without them.
        Row(Modifier.fillMaxWidth()) {
            tickDays.forEachIndexed { i, day ->
                Text(
                    start?.plusDays((day - 1).toLong())?.format(DAY_MONTH) ?: "",
                    Modifier.weight(1f),
                    fontSize = 8.sp,
                    lineHeight = 10.sp,
                    color = Theme.inkSoft.copy(alpha = 0.8f),
                    maxLines = 1,
                    textAlign = when (i) {
                        0 -> TextAlign.Start
                        tickDays.lastIndex -> TextAlign.End
                        else -> TextAlign.Center
                    },
                )
            }
        }
    }
}

// ------------------------------------------------------------------------------------
// Moon
// ------------------------------------------------------------------------------------

private const val MOON_BANDS = 8

@Composable
fun MoonMoodChart(page: CycleStore.MoodPage, modifier: Modifier = Modifier) {
    Canvas(modifier.clearAndSetSemantics {}) {
        val w = size.width - Plot.PAD_X.toPx() * 2
        val h = size.height - Plot.PAD_TOP.toPx() - Plot.PAD_BOTTOM.toPx()
        val bandW = w / MOON_BANDS

        // Lit like the month itself: darkest at the new moon, brightest at full.
        for (i in 0 until MOON_BANDS) {
            val centre = (i + 0.5) / MOON_BANDS
            val illum = (1 - cos(2 * Math.PI * centre)) / 2
            drawRoundRect(
                color = Theme.accent.copy(alpha = (0.06 + 0.16 * illum).toFloat()),
                topLeft = Offset(Plot.PAD_X.toPx() + i * bandW, Plot.PAD_TOP.toPx()),
                size = Size(bandW - 1f, h),
                cornerRadius = CornerRadius(4.dp.toPx(), 4.dp.toPx()),
            )
        }

        with(Plot) {
            // Within one page every dot belongs to the same lunation, so lunar order is date
            // order — the line means exactly what it means in the daily view.
            val points = page.moon.map {
                Offset(Plot.PAD_X.toPx() + it.phaseFraction.toFloat() * w, yOf(it.mood.toDouble())) to it.mood
            }
            connect(points.map { it.first }, Theme.primary.copy(alpha = 0.55f))
            points.forEach { (p, mood) -> moodDot(p, MoodScale.color(mood)) }
        }
    }
}

@Composable
fun MoonMoodAxis(page: CycleStore.MoodPage, modifier: Modifier = Modifier) {
    val store = LocalCycleStore.current
    Row(modifier.fillMaxWidth().padding(horizontal = Plot.PAD_X)) {
        page.moonAverages.forEach { band ->
            val count = band.count
            val spoken = if (count == 0) {
                "${MoonNames.full(band.bucketKey)}, no logs"
            } else {
                val level = MoodScale.label(band.average.toInt().coerceIn(1, 5))
                "${MoonNames.full(band.bucketKey)}, average mood $level, " +
                    "$count log${if (count == 1) "" else "s"}"
            }
            Column(
                Modifier
                    .weight(1f)
                    .semantics { contentDescription = spoken },
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(2.dp),
            ) {
                MoonDisc(
                    illumination = store.moonBucketIllumination(band.bucketKey),
                    waxing = store.moonBucketIsWaxing(band.bucketKey),
                    modifier = Modifier.size(14.dp),
                )
                Text(
                    MoonNames.short(band.bucketKey),
                    fontSize = 7.5.sp,
                    lineHeight = 9.sp,
                    color = Theme.inkSoft,
                    maxLines = 2,
                    textAlign = TextAlign.Center,
                )
            }
        }
    }
}

/** Bucket slugs → display names. Copy lives in the native layer, never in the shared core. */
object MoonNames {
    fun full(key: String): String = when (key) {
        "new" -> "New moon"
        "waxing-crescent" -> "Waxing crescent"
        "first-quarter" -> "First quarter"
        "waxing-gibbous" -> "Waxing gibbous"
        "full" -> "Full moon"
        "waning-gibbous" -> "Waning gibbous"
        "last-quarter" -> "Last quarter"
        else -> "Waning crescent"
    }

    fun short(key: String): String = when (key) {
        "new" -> "New"
        "waxing-crescent" -> "Wax\ncres"
        "first-quarter" -> "First\nqtr"
        "waxing-gibbous" -> "Wax\ngib"
        "full" -> "Full"
        "waning-gibbous" -> "Wan\ngib"
        "last-quarter" -> "Last\nqtr"
        else -> "Wan\ncres"
    }
}

File: iosApp/iosApp/Features/Journal/MoodCharts.swift
import SwiftUI
import Shared

/// The three chart bodies behind "Mood patterns". Each takes one prepared page and draws it —
/// no aggregation here, that all lives in the shared core.
///
/// Shared conventions: y is mood 1..5 (bottom to top), a dot per real log, and a connecting
/// line only once there are enough points for the shape to mean something.
private enum Plot {
    /// Wide enough that a dot on the first or last position isn't clipped by the Canvas
    /// edge. At 6pt a day-1 log was drawn half outside the chart and read as missing.
    static let padX: CGFloat = 12
    static let padTop: CGFloat = 12
    static let padBottom: CGFloat = 10
    /// Below this, joining dots implies a trend the data doesn't show.
    static let minPointsForLine = 4

    static func y(_ mood: Double, _ size: CGSize) -> CGFloat {
        let h = size.height - padTop - padBottom
        return padTop + CGFloat((5 - mood) / 4) * h
    }

    /// One mood dot, ringed in the card colour so overlapping dots stay countable —
    /// consecutive days sit only ~11pt apart on the moon lens, closer than a dot is wide.
    static func dot(_ ctx: GraphicsContext, at p: CGPoint, color: Color) {
        let r: CGFloat = 5
        ctx.fill(Path(ellipseIn: CGRect(x: p.x - r - 1.5, y: p.y - r - 1.5,
                                        width: (r + 1.5) * 2, height: (r + 1.5) * 2)),
                 with: .color(Theme.surface))
        ctx.fill(Path(ellipseIn: CGRect(x: p.x - r, y: p.y - r, width: r * 2, height: r * 2)),
                 with: .color(color))
    }

    static func line(_ pts: [CGPoint]) -> Path {
        var p = Path()
        for (i, pt) in pts.enumerated() { i == 0 ? p.move(to: pt) : p.addLine(to: pt) }
        return p
    }
}

// MARK: - Daily

struct DailyMoodChart: View {
    let page: CycleStore.MoodPage

    var body: some View {
        Canvas { ctx, size in
            let w = size.width - Plot.padX * 2
            let span = max(1, page.spanDays - 1)
            func x(_ dayOffset: Int) -> CGFloat {
                Plot.padX + (CGFloat(dayOffset) / CGFloat(span)) * w
            }

            // Faint gridlines at each mood level, so height is readable without an axis.
            for level in 1...5 {
                let ly = Plot.y(Double(level), size)
                var g = Path()
                g.move(to: CGPoint(x: Plot.padX, y: ly))
                g.addLine(to: CGPoint(x: size.width - Plot.padX, y: ly))
                ctx.stroke(g, with: .color(Theme.inkSoft.opacity(level == 3 ? 0.14 : 0.07)),
                           lineWidth: 1)
            }

            let pts = page.daily.compactMap { d -> CGPoint? in
                guard let offset = dayOffset(d.dateIso) else { return nil }
                return CGPoint(x: x(offset), y: Plot.y(Double(d.mood), size))
            }

            if pts.count >= Plot.minPointsForLine {
                ctx.stroke(Plot.line(pts), with: .color(Theme.primary.opacity(0.55)),
                           style: StrokeStyle(lineWidth: 2.5, lineCap: .round, lineJoin: .round))
            }
            for (i, pt) in pts.enumerated() {
                Plot.dot(ctx, at: pt, color: MoodScale.color(Int(page.daily[i].mood)))
            }
        }
        .accessibilityHidden(true)
    }

    private func dayOffset(_ iso: String) -> Int? {
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd"
        guard let start = f.date(from: page.startIso), let d = f.date(from: iso) else { return nil }
        return Calendar.current.dateComponents([.day], from: start, to: d).day
    }
}

struct DailyMoodAxis: View {
    let page: CycleStore.MoodPage

    /// The phase axis's tick density — enough to place any dot in the month, so the page
    /// needs no separate caption naming its range.
    private static let dateTicks = 5

    var body: some View {
        GeometryReader { geo in
            ZStack(alignment: .topLeading) {
                ForEach(tickOffsets, id: \.self) { offset in
                    let span = CGFloat(max(1, page.spanDays - 1))
                    let x = geo.size.width * CGFloat(offset) / span
                    Text(dateLabel(offset))
                        .font(.system(size: 8))
                        .foregroundStyle(Theme.inkSoft)
                        .frame(width: 44)
                        .offset(x: min(max(0, x - 22), geo.size.width - 44))
                }
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(page.title), \(Int(page.summary.count)) logs")
    }

    private var tickOffsets: [Int] {
        let span = max(1, page.spanDays - 1)
        let step = max(1, span / (Self.dateTicks - 1))
        return Array(stride(from: 0, through: span, by: step))
    }

    private func dateLabel(_ offset: Int) -> String {
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd"
        guard let start = f.date(from: page.startIso),
              let d = Calendar.current.date(byAdding: .day, value: offset, to: start) else { return "" }
        let out = DateFormatter(); out.dateFormat = "d MMM"
        return out.string(from: d)
    }
}

// MARK: - Phase

struct PhaseMoodChart: View {
    let page: CycleStore.MoodPage
    let periodLength: Int

    var body: some View {
        Canvas { ctx, size in
            let w = size.width - Plot.padX * 2
            let h = size.height - Plot.padTop - Plot.padBottom
            let length = max(1, page.spanDays)
            func x(_ day: Int) -> CGFloat {
                Plot.padX + (CGFloat(day - 1) / CGFloat(max(1, length - 1))) * w
            }

            // Phase bands sized to THIS cycle's real length, not the predicted one.
            for phase in PhaseContent.all {
                let r = phase.dayRange(cycleLength: length, periodLength: periodLength)
                let x0 = x(r.lowerBound)
                let x1 = x(min(r.upperBound + 1, length))
                ctx.fill(Path(CGRect(x: x0, y: Plot.padTop, width: max(0, x1 - x0), height: h)),
                         with: .color(phase.color.opacity(0.12)))
            }

            let pts = page.cycle.map {
                CGPoint(x: x(Int($0.cycleDay)), y: Plot.y(Double($0.mood), size))
            }
            if pts.count >= Plot.minPointsForLine {
                ctx.stroke(Plot.line(pts), with: .color(Theme.primary.opacity(0.6)),
                           style: StrokeStyle(lineWidth: 2.5, lineCap: .round, lineJoin: .round))
            }
            for (i, pt) in pts.enumerated() {
                Plot.dot(ctx, at: pt, color: MoodScale.color(Int(page.cycle[i].mood)))
            }
        }
        .accessibilityHidden(true)
    }
}

struct PhaseMoodAxis: View {
    let page: CycleStore.MoodPage
    let periodLength: Int

    /// Roughly this many date ticks across the cycle — enough to locate yourself, few enough
    /// that they don't collide on a narrow screen.
    private static let dateTicks = 5

    var body: some View {
        VStack(spacing: 2) {
            GeometryReader { geo in
                ZStack(alignment: .topLeading) {
                    ForEach(PhaseContent.all, id: \.key) { phase in
                        let r = phase.dayRange(cycleLength: max(1, page.spanDays), periodLength: periodLength)
                        let width = geo.size.width * CGFloat(r.count) / CGFloat(max(1, page.spanDays))
                        let offset = geo.size.width * CGFloat(r.lowerBound - 1) / CGFloat(max(1, page.spanDays))
                        // The ovulatory band is only a few days wide, so its label must
                        // shrink rather than wrap — "Ovulator / y" over two lines was
                        // colliding with the row of dates beneath.
                        Text(phase.key)
                            .font(.system(size: 9)).foregroundStyle(Theme.inkSoft)
                            .lineLimit(1)
                            .minimumScaleFactor(0.6)
                            .frame(width: width)
                            .offset(x: offset)
                    }
                }
            }
            .frame(height: 13)

            // Real calendar dates as well as cycle phases — "day 14" is hard to place in a
            // month without them.
            GeometryReader { geo in
                ZStack(alignment: .topLeading) {
                    ForEach(dateTickDays, id: \.self) { day in
                        let offset = geo.size.width * CGFloat(day - 1) / CGFloat(max(1, page.spanDays))
                        Text(dateLabel(dayOffset: day - 1))
                            .font(.system(size: 8))
                            .foregroundStyle(Theme.inkSoft.opacity(0.8))
                            .frame(width: 44)
                            .offset(x: min(max(0, offset - 22), geo.size.width - 44))
                    }
                }
            }
            .frame(height: 12)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(page.title), \(page.spanDays) days, \(Int(page.summary.count)) logs")
    }

    private var dateTickDays: [Int] {
        let span = max(1, page.spanDays)
        let step = max(1, span / (Self.dateTicks - 1))
        return Array(stride(from: 1, through: span, by: step))
    }

    private func dateLabel(dayOffset: Int) -> String {
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd"
        guard let start = f.date(from: page.startIso),
              let d = Calendar.current.date(byAdding: .day, value: dayOffset, to: start) else { return "" }
        let out = DateFormatter(); out.dateFormat = "d MMM"
        return out.string(from: d)
    }
}

// MARK: - Moon

struct MoonMoodChart: View {
    let page: CycleStore.MoodPage
    private static let bands = 8

    var body: some View {
        Canvas { ctx, size in
            let w = size.width - Plot.padX * 2
            let h = size.height - Plot.padTop - Plot.padBottom
            let bandW = w / CGFloat(Self.bands)

            // Lit like the month itself: darkest at the new moon, brightest at full.
            for i in 0..<Self.bands {
                let centre = (Double(i) + 0.5) / Double(Self.bands)
                let illum = (1 - cos(2 * .pi * centre)) / 2
                let rect = CGRect(x: Plot.padX + CGFloat(i) * bandW, y: Plot.padTop,
                                  width: bandW - 1, height: h)
                ctx.fill(Path(roundedRect: rect, cornerRadius: 4),
                         with: .color(Theme.accent.opacity(0.06 + 0.16 * illum)))
            }

            // Within one page every dot belongs to the same lunation, so lunar order is date
            // order — the line means exactly what it means in the daily view. (Per-band
            // average ticks lived here once and read as floating noise.)
            let pts = page.moon.map {
                CGPoint(x: Plot.padX + CGFloat($0.phaseFraction) * w, y: Plot.y(Double($0.mood), size))
            }
            if pts.count >= Plot.minPointsForLine {
                ctx.stroke(Plot.line(pts), with: .color(Theme.primary.opacity(0.55)),
                           style: StrokeStyle(lineWidth: 2.5, lineCap: .round, lineJoin: .round))
            }
            for (i, pt) in pts.enumerated() {
                Plot.dot(ctx, at: pt, color: MoodScale.color(Int(page.moon[i].mood)))
            }
        }
        .accessibilityHidden(true)
    }
}

struct MoonMoodAxis: View {
    let page: CycleStore.MoodPage
    let store: CycleStore

    var body: some View {
        HStack(spacing: 0) {
            ForEach(Array(page.moonAverages.enumerated()), id: \.element.bucketKey) { index, band in
                VStack(spacing: 2) {
                    MoonDisc(illumination: store.moonBucketIllumination(band.bucketKey),
                             waxing: store.moonBucketIsWaxing(band.bucketKey),
                             glow: false)
                        .frame(width: 14, height: 14)
                    // Dates only, under the principal phases — the discs already picture
                    // the stages, so naming them was noise, and the dates are what place
                    // the lunation in the month (so the page needs no caption). VoiceOver
                    // still names every bucket in full.
                    if let date = bucketDate(index) {
                        Text(date)
                            .font(.system(size: 8))
                            .foregroundStyle(Theme.inkSoft.opacity(0.8))
                            .lineLimit(1).minimumScaleFactor(0.6)
                    }
                }
                .frame(maxWidth: .infinity)
                .accessibilityElement(children: .ignore)
                .accessibilityLabel(description(band))
            }
        }
        .padding(.horizontal, Plot.padX)
    }

    /// The day this bucket begins, for every other bucket (new, first quarter, full, last
    /// quarter). Within one lunation the buckets are date-ordered eighths of its span.
    private func bucketDate(_ index: Int) -> String? {
        guard index % 2 == 0 else { return nil }
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd"
        guard let start = f.date(from: page.startIso),
              let d = Calendar.current.date(byAdding: .day, value: index * page.spanDays / 8, to: start)
        else { return nil }
        let out = DateFormatter(); out.dateFormat = "d MMM"
        return out.string(from: d)
    }

    private func description(_ band: MoonMood) -> String {
        let count = Int(band.count)
        guard count > 0 else { return "\(MoonNames.full(band.bucketKey)), no logs" }
        let level = MoodScale.label(max(1, min(5, Int(band.average.rounded()))))
        return "\(MoonNames.full(band.bucketKey)), average mood \(level), \(count) log\(count == 1 ? "" : "s")"
    }
}

/// Bucket slugs → display names. Copy lives in the native layer, never in the shared core.
enum MoonNames {
    static func full(_ key: String) -> String {
        switch key {
        case "new":             return "New moon"
        case "waxing-crescent": return "Waxing crescent"
        case "first-quarter":   return "First quarter"
        case "waxing-gibbous":  return "Waxing gibbous"
        case "full":            return "Full moon"
        case "waning-gibbous":  return "Waning gibbous"
        case "last-quarter":    return "Last quarter"
        default:                return "Waning crescent"
        }
    }

}
